utils/symbol.c

```diff
@@ -1430,3 +1423,61 @@ void print_symtabs(struct symtabs *symtabs)
 		symbol_putname(sym, name);
 	}
 }
+
+static unsigned long kernel_base_addr;
+
+static unsigned long get_kernel_base(char *str)
+{
+	unsigned long addr = strtoul(str, NULL, 16);
+
+	if (addr < 0x40000000UL) {
+		return 0x40000000UL;
+	} else if (addr < 0x80000000UL) {
+		return 0x80000000UL;
+	} else if (addr < 0xB0000000UL) {
+		return 0xB0000000UL;
+	} else if (addr < 0xC0000000UL) {
+		return 0xC0000000UL;
+	} else {
+		return 0x800000000000ULL;
+	}
+}
+
+void set_kernel_base(char *dirname, const char *session_id)
+{
+	FILE *fp;
+	char buf[4096];
+	char line[200];
+
+	snprintf(buf, sizeof(buf), "%s/sid-%s.map", dirname, session_id);
+
+	fp = fopen(buf, "r");
+	if (fp == NULL) {
+		if (errno == EEXIST)
+			return;
+		pr_err("cannot open %s file", buf);
+	}
+
+	while (fgets(line, sizeof(line), fp) != NULL) {
+		if (strstr(line, "[stack") != NULL) {
+			kernel_base_addr = get_kernel_base(line);
+			break;
+		}
+	}
+	fclose(fp);
+}
+
+bool is_kernel_address(unsigned long addr)
+{
+	return addr >= kernel_base_addr;
+}
+
+unsigned long get_real_address(unsigned long addr)
+{
+	if (is_kernel_address(addr)) {
+		return (kernel_base_addr == 0x800000000000ULL) ?
+                    (addr | (-1UL << KADDR_SHIFT)) :
+                    (addr | kernel_base_addr);
+	}
+	return addr;
+}
```

Comment:
@namhyung, your suggestion looks better, but I am still thinking about it.

As you mentioned, replay can be run on different machine. IMHO, It means that uftrace can replay any uftrace.data where it records from; e.g., the uftrace.data is recorded on arm32 and it can replay on x86_64.

Applying your suggestion, `uftrace replay --kernel` command can replay the uftrace.data from 64-bit machine only. I am attaching a tracing result on 64-bit machine with 32-bit uftrace.data:

```
uftrace replay --kernel
# DURATION    TID     FUNCTION
   4.000 us [ 2456] | __cxa_atexit();
   6.300 us [ 2456] | <4008002c>();
   4.000 us [ 2456] | main();
            [ 2456] | <4008002c>() {
            [ 2456] |   printf() {
   3.900 us [ 2456] |   } /* <4008002c> */
  13.400 us [ 2456] |   <40112108>();
            [ 2456] |   <4011df44>() {
  16.800 us [ 2456] |     <4000a3f4>();
  61.200 us [ 2456] |   } /* <4011df44> */
  34.500 us [ 2456] |   <400f5f2c>();
  33.800 us [ 2456] |   <4001efac>();
  81.900 us [ 2456] |   <4010e1a4>();
            [ 2456] |   <4008002c>() {
   6.400 us [ 2456] |   } /* printf */
 295.900 us [ 2456] | } /* <4008002c> */
   6.200 us [ 2456] | <4008002c>();

uftrace stopped tracing with remaining functions
================================================
task: 2456
[0] <4002cfa8>
```

The reason why I wrote code as below:
```
unsigned long get_real_address(unsigned long addr)
{
       if (is_kernel_address(addr)) {
               return (kernel_base_addr == 0x800000000000ULL) ?
                    (addr | (-1UL << KADDR_SHIFT)) :
                    (addr | kernel_base_addr);
        return addr;
}
```